evm/src/keccak/keccak_stark.rs

```diff
@@ -1,60 +1,522 @@
+
+/// Number of 64-bit limbs in a preimage of the Keccak permutation.
+pub(crate) const INPUT_LIMBS: usize = 25;
+
+pub(crate) const NUM_PUBLIC_INPUTS: usize = 0;
+
 #[derive(Copy, Clone)]
 pub struct KeccakStark<F, const D: usize> {
     pub(crate) f: PhantomData<F>,
 }
 
+impl<F: RichField + Extendable<D>, const D: usize> KeccakStark<F, D> {
+    /// Generate the rows of the trace. Note that this does not generate the permuted columns used
+    /// in our lookup arguments, as those are computed after transposing to column-wise form.
+    pub(crate) fn generate_trace_rows(
+        &self,
+        inputs: Vec<[u64; INPUT_LIMBS]>,
+    ) -> Vec<[F; NUM_REGISTERS]> {
+        let num_rows = (inputs.len() * NUM_ROUNDS).next_power_of_two();
+        info!("{} rows", num_rows);
+        let mut rows = Vec::with_capacity(num_rows);
+        for input in inputs.iter().take(1) {
+            rows.extend(self.generate_trace_rows_for_perm(*input));
+        }
+
+        // Pad rows to power of two.
+        for i in rows.len()..num_rows {
+            let mut row = [F::ZERO; NUM_REGISTERS];
+            self.copy_output_to_input(rows[i - 1], &mut row);
+            self.generate_trace_rows_for_round(&mut row, i % NUM_ROUNDS);
+            rows.push(row);
+        }
+
+        rows
+    }
+
+    fn generate_trace_rows_for_perm(&self, input: [u64; INPUT_LIMBS]) -> Vec<[F; NUM_REGISTERS]> {
+        let mut rows = vec![[F::ZERO; NUM_REGISTERS]; NUM_ROUNDS];
+
+        for x in 0..5 {
+            for y in 0..5 {
+                let input_xy = input[x * 5 + y];
+                for z in 0..64 {
+                    rows[0][reg_a(x, y, z)] = F::from_canonical_u64((input_xy >> z) & 1);
+                }
+            }
+        }
+
+        self.generate_trace_rows_for_round(&mut rows[0], 0);
+        for round in 1..24 {
+            self.copy_output_to_input(rows[round - 1], &mut rows[round]);
+            self.generate_trace_rows_for_round(&mut rows[round], round);
+        }
+
+        rows
+    }
+
+    fn copy_output_to_input(
+        &self,
+        prev_row: [F; NUM_REGISTERS],
+        next_row: &mut [F; NUM_REGISTERS],
+    ) {
+        for x in 0..5 {
+            for y in 0..5 {
+                let cur_lo = prev_row[reg_a_prime_prime_prime(x, y)];
+                let cur_hi = prev_row[reg_a_prime_prime_prime(x, y) + 1];
+                let cur_u64 = cur_lo.to_canonical_u64() + (1 << 32) * cur_hi.to_canonical_u64();
+                let bit_values: Vec<u64> = (0..64)
+                    .scan(cur_u64, |acc, _| {
+                        let tmp = *acc % 2;
+                        *acc /= 2;
+                        Some(tmp)
+                    })
+                    .collect();
+
+                for z in 0..64 {
+                    next_row[reg_a(x, y, z)] = F::from_canonical_u64(bit_values[z]);
+                }
+            }
+        }
+    }
+
+    fn generate_trace_rows_for_round(&self, row: &mut [F; NUM_REGISTERS], round: usize) {
```

Comment:
nit: seems like it should be singular "row" (sorry if it's my typo)

Reply:
Ah yes